File: backend/engine/roughness.py
"""ISO 4287 / ISO 21920 / ASME B46.1 surface roughness parameter calculations.

All inputs are height arrays in µm sampled at uniform lateral spacing dx (µm).
The engine is intentionally pure — no FastAPI or HTTP dependencies — so the same
code can later be called from an analyzer endpoint that ingests real measurement data.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np


@dataclass
class RoughnessParameters:
    Ra: float   # Arithmetic mean deviation
    Rq: float   # RMS deviation
    Rp: float   # Maximum peak height
    Rv: float   # Maximum valley depth (positive value)
    Rz: float   # Mean roughness depth (5-segment average)
    Rt: float   # Total height
    Rsk: float  # Skewness (dimensionless)
    Rku: float  # Kurtosis (dimensionless)
    Rc: float   # Mean height of profile elements
    RSm: float  # Mean width of profile elements (µm)
    RDq: float  # RMS profile slope (degrees)
    Rk: float   # Core roughness depth (ISO 13565-2)
    Rpk: float  # Reduced peak height (ISO 13565-2)
    Rvk: float  # Reduced valley depth (ISO 13565-2)


@dataclass
class MaterialRatioPoint:
    c: float    # Depth below highest peak (µm)
    mr: float   # Material ratio (0–100 %)


def center_profile(z: np.ndarray) -> np.ndarray:
    """Subtract the mean (centre) line from a profile array."""
    return z - z.mean()


def _compute_rz(z: np.ndarray, segments: int = 5) -> float:
    """Mean roughness depth: average peak-to-valley over `segments` sampling lengths."""
    n = len(z)
    if n == 0:
        return 0.0
    seg_len = n // segments
    if seg_len < 2:
        return float(z.max() - z.min())
    total = 0.0
    for s in range(segments):
        start = s * seg_len
        end = start + seg_len if s < segments - 1 else n
        seg = z[start:end]
        total += float(seg.max() - seg.min())
    return total / segments


def _compute_element_stats(z: np.ndarray, dx: float, rq: float) -> tuple[float, float]:
    """Compute RSm (mean element spacing) and Rc (mean element height).

    Profile elements are the segments between consecutive upward mean-line
    crossings. A height discrimination of 0.1·Rq suppresses spurious crossings
    from fine noise (spirit of ISO 4287 §4).
    """
    n = len(z)
    threshold = rq * 0.1
    crossings: list[int] = []
    for i in range(1, n):
        if z[i - 1] <= 0 and z[i] > threshold:
            crossings.append(i)

    if len(crossings) < 2:
        height = float(z.max() - z.min())
        return float(n * dx), height

    width_sum = 0.0
    height_sum = 0.0
    count = len(crossings) - 1
    for k in range(1, len(crossings)):
        start = crossings[k - 1]
        end = crossings[k]
        width_sum += (end - start) * dx
        seg = z[start:end]
        height_sum += float(seg.max() - seg.min())

    return width_sum / count, height_sum / count


def _compute_rdq(z: np.ndarray, dx: float) -> float:
    """RMS profile slope in degrees."""
    if len(z) < 2 or dx == 0:
        return 0.0
    slopes = np.diff(z) / dx  # dimensionless (µm/µm)
    rms_slope = float(np.sqrt(np.mean(slopes ** 2)))
    return math.degrees(math.atan(rms_slope))


@dataclass
class CoreRoughnessParameters:
    Rk: float
    Rpk: float
    Rvk: float
    Mr1: float  # Material ratio (%) bounding the peak zone
    Mr2: float  # Material ratio (%) bounding the valley zone


def _compute_core_roughness(z: np.ndarray) -> CoreRoughnessParameters:
    """ISO 13565-2 core roughness parameters from the linearized bearing-area curve.

    Slides a secant spanning 40% of the material ratio across the sorted
    (bearing-area) profile to find its flattest span — the core — then extends
    that line to material ratios 0% and 100%. Rk is the vertical span between
    those two extensions; Rpk and Rvk are the heights of right triangles whose
    areas match the true peak/valley material lying outside the core band.
    """
    n = len(z)
    if n < 3:
        return CoreRoughnessParameters(Rk=0.0, Rpk=0.0, Rvk=0.0, Mr1=0.0, Mr2=100.0)

    zs = np.sort(z)[::-1]  # descending: index position ↔ material ratio
    window = min(n - 1, max(1, round(0.4 * n)))

    diffs = zs[:-window] - zs[window:]
    i0 = int(np.argmin(diffs))
    slope = (zs[i0 + window] - zs[i0]) / window

    h0 = float(zs[i0] - slope * i0)            # core line extended to mr = 0%
    h100 = float(zs[i0] + slope * (n - i0))     # core line extended to mr = 100%
    Rk = h0 - h100

    mr1_idx = int(np.searchsorted(-zs, -h0, side="left"))
    mr2_idx = int(np.searchsorted(-zs, -h100, side="left"))
    mr1_idx = min(max(mr1_idx, 1), n - 1)
    mr2_idx = min(max(mr2_idx, mr1_idx), n - 1)

    peak_excess = float(np.sum(zs[:mr1_idx] - h0))
    Rpk = max(0.0, 2 * peak_excess / mr1_idx)

    tail = n - mr2_idx
    valley_excess = float(np.sum(h100 - zs[mr2_idx:]))
    Rvk = max(0.0, 2 * valley_excess / tail) if tail > 0 else 0.0

    return CoreRoughnessParameters(
        Rk=Rk, Rpk=Rpk, Rvk=Rvk,
        Mr1=mr1_idx / n * 100.0, Mr2=mr2_idx / n * 100.0,
    )


def compute_parameters(z: np.ndarray, dx: float) -> RoughnessParameters:
    """Compute the full ISO 4287 2D parameter suite from a centred profile."""
    z = center_profile(z)
    n = len(z)
    if n == 0:
        return RoughnessParameters(
            Ra=0, Rq=0, Rp=0, Rv=0, Rz=0, Rt=0,
            Rsk=0, Rku=0, Rc=0, RSm=0, RDq=0,
            Rk=0, Rpk=0, Rvk=0,
        )

    Ra = float(np.mean(np.abs(z)))
    Rq = float(np.sqrt(np.mean(z ** 2)))
    Rp = float(z.max())
    Rv = float(-z.min())
    Rt = Rp + Rv
    Rz = _compute_rz(z)
    Rsk = float(np.mean(z ** 3) / Rq ** 3) if Rq > 0 else 0.0
    Rku = float(np.mean(z ** 4) / Rq ** 4) if Rq > 0 else 0.0
    RSm, Rc = _compute_element_stats(z, dx, Rq)
    RDq = _compute_rdq(z, dx)
    core = _compute_core_roughness(z)

    return RoughnessParameters(
        Ra=Ra, Rq=Rq, Rp=Rp, Rv=Rv, Rz=Rz, Rt=Rt,
        Rsk=Rsk, Rku=Rku, Rc=Rc, RSm=RSm, RDq=RDq,
        Rk=core.Rk, Rpk=core.Rpk, Rvk=core.Rvk,
    )


def material_ratio_curve(z: np.ndarray, steps: int = 100) -> list[MaterialRatioPoint]:
    """Abbott–Firestone bearing area (material ratio) curve."""
    z = center_profile(z)
    n = len(z)
    hi = float(z.max())
    lo = float(z.min())
    span = hi - lo
    if span == 0:
        return [MaterialRatioPoint(c=0.0, mr=100.0)]

    sorted_z = np.sort(z)[::-1]  # descending
    points: list[MaterialRatioPoint] = []
    idx = 0
    for s in range(steps + 1):
        level = hi - (span * s / steps)
        while idx < n and sorted_z[idx] >= level:
            idx += 1
        points.append(MaterialRatioPoint(c=hi - level, mr=(idx / n) * 100.0))
    return points

File: src/lib/roughness.ts
// Surface roughness parameter calculations (2D profile / R-parameters).
//
// Definitions follow ISO 4287 / ISO 21920-2 (and ASME B46.1, which agrees on
// the core amplitude parameters). All amplitude/height values are returned in
// micrometres (µm); the lateral spacing parameter RSm is returned in
// micrometres of lateral distance; RΔq is returned in degrees; Rsk and Rku are
// dimensionless.
//
// The engine is intentionally pure and decoupled from how a profile is created,
// so the same code can later compute parameters from imported measurement data
// (the planned "analyzer" feature), not just synthesized traces.

/** A sampled surface profile. */
export interface Profile {
  /** Heights in µm, sampled at uniform spacing. Need not be mean-centred. */
  z: number[];
  /** Lateral spacing between samples, in µm. */
  dx: number;
}

export interface MaterialRatioPoint {
  /** Depth below the highest peak, in µm. */
  c: number;
  /** Material ratio at that depth, in percent (0–100). */
  mr: number;
}

export interface RoughnessParameters {
  /** Arithmetic mean deviation. */
  Ra: number;
  /** Root-mean-square (RMS) deviation. */
  Rq: number;
  /** Maximum peak height above the mean line. */
  Rp: number;
  /** Maximum valley depth below the mean line (positive). */
  Rv: number;
  /** Mean roughness depth (mean peak-to-valley over sampling lengths). */
  Rz: number;
  /** Total height of the profile (max peak + max valley). */
  Rt: number;
  /** Skewness — asymmetry of the height distribution (dimensionless). */
  Rsk: number;
  /** Kurtosis — sharpness/spikiness of the height distribution (dimensionless). */
  Rku: number;
  /** Mean height of profile elements. */
  Rc: number;
  /** Mean width of profile elements (lateral), in µm. */
  RSm: number;
  /** Root-mean-square profile slope, in degrees. */
  RDq: number;
}

function mean(values: number[]): number {
  let sum = 0;
  for (const v of values) sum += v;
  return sum / values.length;
}

/** Returns a copy of the profile with its mean (centre) line subtracted. */
export function centerProfile(z: number[]): number[] {
  const m = mean(z);
  return z.map((v) => v - m);
}

/**
 * Mean roughness depth Rz (ISO): split the evaluation length into `segments`
 * equal sampling lengths, take the peak-to-valley height in each, and average.
 */
function computeRz(z: number[], segments = 5): number {
  const n = z.length;
  if (n === 0) return 0;
  const len = Math.floor(n / segments);
  if (len < 2) {
    return Math.max(...z) - Math.min(...z);
  }
  let total = 0;
  for (let s = 0; s < segments; s++) {
    const start = s * len;
    const end = s === segments - 1 ? n : start + len;
    let hi = -Infinity;
    let lo = Infinity;
    for (let i = start; i < end; i++) {
      if (z[i] > hi) hi = z[i];
      if (z[i] < lo) lo = z[i];
    }
    total += hi - lo;
  }
  return total / segments;
}

/**
 * Profile elements are the segments between consecutive upward crossings of the
 * mean line. A small height discrimination (a fraction of Rq) suppresses
 * spurious crossings from fine noise, per the spirit of ISO discrimination.
 * Returns mean element width (µm) for RSm and mean element height (µm) for Rc.
 */
function computeElementStats(
  z: number[],
  dx: number,
  rq: number,
): { RSm: number; Rc: number } {
  const n = z.length;
  const threshold = rq * 0.1; // height discrimination
  const upwardCrossings: number[] = [];
  for (let i = 1; i < n; i++) {
    if (z[i - 1] <= 0 && z[i] > threshold) {
      upwardCrossings.push(i);
    }
  }
  if (upwardCrossings.length < 2) {
    return { RSm: n * dx, Rc: Math.max(...z) - Math.min(...z) };
  }
  let widthSum = 0;
  let heightSum = 0;
  let count = 0;
  for (let k = 1; k < upwardCrossings.length; k++) {
    const start = upwardCrossings[k - 1];
    const end = upwardCrossings[k];
    widthSum += (end - start) * dx;
    let hi = -Infinity;
    let lo = Infinity;
    for (let i = start; i < end; i++) {
      if (z[i] > hi) hi = z[i];
      if (z[i] < lo) lo = z[i];
    }
    heightSum += hi - lo;
    count++;
  }
  return { RSm: widthSum / count, Rc: heightSum / count };
}

/** RMS profile slope RΔq, returned in degrees. */
function computeRDq(z: number[], dx: number): number {
  if (z.length < 2 || dx === 0) return 0;
  let sumSq = 0;
  for (let i = 1; i < z.length; i++) {
    const slope = (z[i] - z[i - 1]) / dx; // µm / µm = dimensionless ratio
    sumSq += slope * slope;
  }
  const rmsSlope = Math.sqrt(sumSq / (z.length - 1));
  return (Math.atan(rmsSlope) * 180) / Math.PI;
}

/**
 * Abbott–Firestone material ratio (bearing ratio) curve. For each depth `c`
 * below the highest peak, the material ratio is the fraction of samples at or
 * above that height.
 */
export function materialRatioCurve(
  profile: Profile,
  steps = 100,
): MaterialRatioPoint[] {
  const z = centerProfile(profile.z);
  const n = z.length;
  const hi = Math.max(...z);
  const lo = Math.min(...z);
  const range = hi - lo;
  if (range === 0) return [{ c: 0, mr: 100 }];
  const sorted = [...z].sort((a, b) => b - a); // descending
  const points: MaterialRatioPoint[] = [];
  let idx = 0;
  for (let s = 0; s <= steps; s++) {
    const level = hi - (range * s) / steps; // walk down from the highest peak
    while (idx < n && sorted[idx] >= level) idx++;
    points.push({ c: hi - level, mr: (idx / n) * 100 });
  }
  return points;
}

/** Compute the full standard 2D roughness parameter suite for a profile. */
export function computeParameters(profile: Profile): RoughnessParameters {
  const z = centerProfile(profile.z);
  const n = z.length;
  if (n === 0) {
    return {
      Ra: 0, Rq: 0, Rp: 0, Rv: 0, Rz: 0, Rt: 0,
      Rsk: 0, Rku: 0, Rc: 0, RSm: 0, RDq: 0,
    };
  }

  let absSum = 0;
  let sq = 0;
  let cube = 0;
  let quad = 0;
  let hi = -Infinity;
  let lo = Infinity;
  for (const v of z) {
    absSum += Math.abs(v);
    const v2 = v * v;
    sq += v2;
    cube += v2 * v;
    quad += v2 * v2;
    if (v > hi) hi = v;
    if (v < lo) lo = v;
  }

  const Ra = absSum / n;
  const Rq = Math.sqrt(sq / n);
  const Rp = hi;
  const Rv = -lo;
  const Rt = Rp + Rv;
  const Rz = computeRz(z);
  const Rsk = Rq > 0 ? cube / n / (Rq * Rq * Rq) : 0;
  const Rku = Rq > 0 ? quad / n / (Rq * Rq * Rq * Rq) : 0;
  const { RSm, Rc } = computeElementStats(z, profile.dx, Rq);
  const RDq = computeRDq(z, profile.dx);

  return { Ra, Rq, Rp, Rv, Rz, Rt, Rsk, Rku, Rc, RSm, RDq };
}
